Add error location in higher layers

vijayphoenix opened the pull request:
1. These changes introduce error srcLoc which help in showing better error msgs at a higher level
2. Minor code improvements
3. These changes will not become a source for some other errors/problems.
Stats on YAML-test suite

```
done -- passed: 316 (ev: 32, ev+json: 93, ev+json+yaml: 120, err: 71) / failed: 2 (err: 2, ev:0, json:0, yaml:0, ok:0)
```
More commits will follow to improve error loc in YAML.hs

hvr commented on src/Data/YAML/Loader.hs:
```diff
@@ -46,12 +46,12 @@ data Loader m n = Loader
-        | YE.posCharOffset pos < 0 -> return (Left err)
-        | otherwise                -> return (Left $ ":" ++ show (YE.posLine pos) ++ ":" ++ show (YE.posColumn pos) ++ ": " ++ err)
+        | YE.posCharOffset pos < 0 -> return (Left (pos,err))
+        | otherwise                -> return (Left (pos,":" ++ show (YE.posLine pos) ++ ":" ++ show (YE.posColumn pos) ++ ": " ++ err))
```
btw, we don't need to include the line/column position in the error string anymore

vijayphoenix commented:
1. Removed rotatePosList to improve efficiency
2. Removed line/column position in the error string https://github.com/haskell-hvr/HsYAML/pull/19#discussion_r294055940
3. Added error SrcLoc in YAML.hs
4. Custom (Eq, Ord) instance for Node in YAML.hs
Stats on YAML-test suite
```
done -- passed: 316 (ev: 32, ev+json: 93, ev+json+yaml: 120, err: 71) / failed: 2 (err: 2, ev:0, json:0, yaml:0, ok:0)
```
